File: fake_generator.cpp
#include "fake_generator.h"
#include <iostream>
#include <sstream>
#include <iomanip>
#include <random>
#include <map>
#include <algorithm>
#include <ctime>

namespace fake_generator {
    
    // Random number generator
    std::random_device rd;
    std::mt19937 gen(rd());
    
    // Character sets for password generation
    const std::string lowercase = "abcdefghijklmnopqrstuvwxyz";
    const std::string uppercase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    const std::string digits = "0123456789";
    const std::string special = "!@#$%^&*";
    
    // Sample names for random generation
    const std::vector<std::string> firstNames = {
        "James", "John", "Robert", "Michael", "William", 
        "David", "Richard", "Joseph", "Thomas", "Charles",
        "Mary", "Patricia", "Jennifer", "Linda", "Barbara",
        "Elizabeth", "Susan", "Jessica", "Sarah", "Karen"
    };
    
    const std::vector<std::string> lastNames = {
        "Smith", "Johnson", "Williams", "Brown", "Jones",
        "Garcia", "Miller", "Davis", "Rodriguez", "Martinez",
        "Hernandez", "Lopez", "Gonzalez", "Wilson", "Anderson"
    };
    
    // Email domains
    const std::vector<std::string> emailDomains = {
        "gmail.com", "yahoo.com", "outlook.com", 
        "hotmail.com", "mail.com", "example.com"
    };
    
    // Dictionary of realistic single words for honey encryption
    const std::vector<std::string> wordDictionary = {
        "account", "admin", "balance", "bank", "billing", "budget", "card",
        "cash", "client", "cloud", "code", "config", "credit", "crypto",
        "customer", "data", "database", "domain", "download", "driver",
        "email", "encryption", "file", "folder", "gateway", "hardware",
        "host", "image", "invoice", "key", "license", "login", "mail",
        "network", "note", "order", "packet", "password", "payment", "phone",
        "portal", "private", "profile", "project", "proxy", "public",
        "query", "record", "report", "request", "resource", "router",
        "script", "secure", "server", "service", "session", "share",
        "software", "source", "storage", "stream", "system", "token",
        "traffic", "transaction", "upload", "user", "video", "voice",
        "wallet", "web", "wireless", "worker", "zone"
    };
    
    // Markov chain training corpus
    const std::string corpus = 
        "The quick brown fox jumps over the lazy dog. "
        "Security is paramount in modern computing systems. "
        "Data protection requires strong encryption methods. "
        "Network security prevents unauthorized access. "
        "Authentication verifies user identity. "
        "Passwords should be strong and unique. "
        "Firewalls protect against malicious attacks. "
        "Encryption ensures data confidentiality. "
        "Digital signatures provide authenticity. "
        "Intrusion detection systems monitor network traffic. "
        "Malware can compromise system integrity. "
        "Backup strategies prevent data loss. "
        "Access control limits system resources. "
        "VPN creates secure remote connections. "
        "Cyber security protects digital assets.";
    
    // Markov chain model
    std::map<char, std::vector<char>> markovModel;
    
    /**
     * @brief Build Markov chain model from corpus
     */
    void buildMarkovModel() {
        if (!markovModel.empty()) return;
        
        for (size_t i = 0; i < corpus.length() - 1; ++i) {
            char c = corpus[i];
            char next = corpus[i + 1];
            if (c != ' ') {
                markovModel[c].push_back(next);
            }
        }
    }
    
    std::string generateRandomName() {
        std::uniform_int_distribution<> firstDist(0, firstNames.size() - 1);
        std::uniform_int_distribution<> lastDist(0, lastNames.size() - 1);
        
        return firstNames[firstDist(gen)] + " " + lastNames[lastDist(gen)];
    }
    
    std::string generateRandomPassword(size_t length) {
        std::string allChars = lowercase + uppercase + digits + special;
        std::uniform_int_distribution<> dist(0, allChars.length() - 1);
        
        std::string password;
        password.reserve(length);
        
        // Ensure at least one of each type
        password += lowercase[std::uniform_int_distribution<>(0, lowercase.length() - 1)(gen)];
        password += uppercase[std::uniform_int_distribution<>(0, uppercase.length() - 1)(gen)];
        password += digits[std::uniform_int_distribution<>(0, digits.length() - 1)(gen)];
        password += special[std::uniform_int_distribution<>(0, special.length() - 1)(gen)];
        
        // Fill the rest randomly
        for (size_t i = 4; i < length; ++i) {
            password += allChars[dist(gen)];
        }
        
        // Shuffle the password
        std::shuffle(password.begin(), password.end(), gen);
        
        return password;
    }
    
    std::string generateRandomEmail() {
        std::uniform_int_distribution<> firstDist(0, firstNames.size() - 1);
        std::uniform_int_distribution<> lastDist(0, lastNames.size() - 1);
        std::uniform_int_distribution<> domainDist(0, emailDomains.size() - 1);
        std::uniform_int_distribution<> numDist(100, 9999);
        
        std::string email(1, lowercase[std::uniform_int_distribution<>(0, lowercase.length() - 1)(gen)]);
        email += lastNames[lastDist(gen)];
        email += std::to_string(numDist(gen));
        email += "@";
        email += emailDomains[domainDist(gen)];
        
        return email;
    }
    
    std::string generateFakeCredentials() {
        std::string name = generateRandomName();
        std::string password = generateRandomPassword(12);
        std::string email = generateRandomEmail();
        
        std::ostringstream oss;
        oss << "USERNAME: " << name << "\n";
        oss << "PASSWORD: " << password << "\n";
        oss << "EMAIL: " << email;
        
        return oss.str();
    }
    
    std::string generateMarkovText(size_t length) {
        buildMarkovModel();
        
        if (markovModel.empty()) {
            return "Sample text generated by Markov chain model.";
        }
        
        std::string result;
        result.reserve(length);
        
        // Start with a random letter
        std::uniform_int_distribution<> letterDist(0, 25);
        char current = 'a' + letterDist(gen);
        result += current;
        
        while (result.length() < length) {
            auto it = markovModel.find(current);
            if (it == markovModel.end() || it->second.empty()) {
                // If no continuation, pick a random letter
                current = 'a' + letterDist(gen);
            } else {
                // Pick next character from model
                std::uniform_int_distribution<> 
                    nextDist(0, it->second.size() - 1);
                current = it->second[nextDist(gen)];
            }
            
            result += current;
            
            // Add spaces occasionally
            if (current == '.' || current == ' ' || 
                std::uniform_int_distribution<>(0, 20)(gen) == 0) {
                result += ' ';
            }
        }
        
        // Make it look like sentences
        if (result.length() > 10) {
            result[0] = std::toupper(result[0]);
            // Add proper sentence ending
            if (result[result.length() - 1] != '.') {
                result += ".";
            }
        }
        
        return result;
    }
    
    std::string generateFakePlaintext() {
        // Randomly choose between the two methods
        std::uniform_int_distribution<> choiceDist(0, 1);
        int choice = choiceDist(gen);
        
        if (choice == 0) {
            // Frequency-based model (credential format)
            return generateFakeCredentials();
        } else {
            // Markov chain generator
            // Generate text of similar length to typical credentials
            std::uniform_int_distribution<> lengthDist(40, 80);
            return generateMarkovText(lengthDist(gen));
        }
    }
    
    std::string generateFakeWord() {
        // Randomly select a word from the dictionary
        std::uniform_int_distribution<> dist(0, wordDictionary.size() - 1);
        return wordDictionary[dist(gen)];
    }
}

File: backend/crypto/fake_generator.py
"""
Fake Plaintext Generator

Generates realistic-looking fake plaintexts for honey encryption:
- Chat messages
- Credentials (username/password/email)
- Email format messages
"""

import random
import string


class FakeGenerator:
    """Generates fake plaintexts for honey encryption"""
    
    def __init__(self):
        self._init_data()
    
    def _init_data(self):
        """Initialize data lists for fake generation"""
        
        # Chat message templates
        self.chat_templates = [
            "Hey, let's meet {time} {day}.",
            "Did you see the {topic}?",
            "Can you send me the {item}?",
            "The meeting is at {time}.",
            "I'll send you the {item} later.",
            "Thanks for the {item}!",
            "Sure, no problem!",
            "Let me check and get back to you.",
            "The {topic} looks great!",
            "I'll be there at {time}.",
            "Don't forget the {item}!",
            "See you {day}!",
            "Great work on the {topic}!",
            "Can you call me later?",
            "Just finished the {item}.",
            "The {item} is ready.",
            "Let's discuss this {day}.",
            "Perfect, that works for me!",
            "I'll {action} tomorrow.",
            "Thanks for your help!"
        ]
        
        # Chat placeholders
        self.times = ["9am", "10am", "2pm", "3pm", "4pm", "5pm", "morning", "afternoon", "evening"]
        self.days = ["today", "tomorrow", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday"]
        self.topics = ["project", "report", "presentation", "meeting", "document", "email", "message", "update"]
        self.items = ["file", "document", "report", "email", "message", "details", "info", "data", "results"]
        self.actions = ["send it", "finish it", "complete it", "review it", "check it", "update it"]
        
        # First names
        self.first_names = [
            "James", "John", "Robert", "Michael", "William", "David", "Richard", 
            "Joseph", "Thomas", "Charles", "Mary", "Patricia", "Jennifer", "Linda", 
            "Barbara", "Elizabeth", "Susan", "Jessica", "Sarah", "Karen", "Lisa",
            "Nancy", "Betty", "Margaret", "Sandra", "Ashley", "Kimberly", "Emily",
            "Donna", "Michelle", "Carol", "Amanda", "Melissa", "Deborah", "Stephanie"
        ]
        
        # Last names
        self.last_names = [
            "Smith", "Johnson", "Williams", "Brown", "Jones", "Garcia", "Miller", 
            "Davis", "Rodriguez", "Martinez", "Hernandez", "Lopez", "Gonzalez", 
            "Wilson", "Anderson", "Thomas", "Taylor", "Moore", "Jackson", "Martin",
            "Lee", "Perez", "Thompson", "White", "Harris", "Sanchez", "Clark"
        ]
        
        # Email domains
        self.email_domains = [
            "gmail.com", "yahoo.com", "outlook.com", "hotmail.com", 
            "mail.com", "example.com", "company.com", "business.org"
        ]
        
        # Email subjects
        self.email_subjects = [
            "Meeting Update", "Project Status", "Quick Question", "Follow Up",
            "Re: Document", "Important", "Information", "Request",
            "Confirmation", "Reminder", "Thanks", "Update on Matter"
        ]
        
        # Email message starters
        self.email_starters = [
            "I wanted to follow up on", "Please find the", "Thank you for your",
            "I hope this email", "Just a quick", "As discussed", 
            "Following up on our", "I am writing to"
        ]
        
        # Single words dictionary
        self.word_dictionary = [
            "account", "admin", "balance", "bank", "billing", "budget", "card",
            "cash", "client", "cloud", "code", "config", "credit", "crypto",
            "customer", "data", "database", "domain", "download", "driver",
            "email", "encryption", "file", "folder", "gateway", "hardware",
            "host", "image", "invoice", "key", "license", "login", "mail",
            "network", "note", "order", "packet", "password", "payment", "phone",
            "portal", "private", "profile", "project", "proxy", "public",
            "query", "record", "report", "request", "resource", "router",
            "script", "secure", "server", "service", "session", "share",
            "software", "source", "storage", "stream", "system", "token",
            "traffic", "transaction", "upload", "user", "video", "voice",
            "wallet", "web", "wireless", "worker", "zone", "ready", "done",
            "sent", "received", "approved", "pending", "waiting", "processing"
        ]
    
    def _generate_random_password(self, length: int = 12) -> str:
        """Generate a random password-like string"""
        chars = string.ascii_lowercase + string.ascii_uppercase + string.digits + "!@#$%"
        return ''.join(random.choice(chars) for _ in range(length))
    
    def _generate_random_username(self) -> str:
        """Generate a random username"""
        first = random.choice(self.first_names).lower()
        last = random.choice(self.last_names).lower()
        num = random.randint(1, 999)
        
        patterns = [
            f"{first}{last}",
            f"{first}.{last}",
            f"{first}_{last}",
            f"{first}{num}",
            f"{last}{first}",
            f"{first[0]}{last}{num}"
        ]
        return random.choice(patterns)
    
    def _generate_random_email(self) -> str:
        """Generate a random email address"""
        username = self._generate_random_username()
        domain = random.choice(self.email_domains)
        return f"{username}@{domain}"
    
    def generate_chat_message(self) -> str:
        """Generate a fake chat message"""
        template = random.choice(self.chat_templates)
        
        replacements = {
            "{time}": random.choice(self.times),
            "{day}": random.choice(self.days),
            "{topic}": random.choice(self.topics),
            "{item}": random.choice(self.items),
            "{action}": random.choice(self.actions)
        }
        
        for key, value in replacements.items():
            template = template.replace(key, value)
        
        return template
    
    def generate_credentials(self) -> str:
        """Generate fake credentials"""
        username = self._generate_random_username()
        password = self._generate_random_password()
        email = self._generate_random_email()
        
        return f"USERNAME: {username}\nPASSWORD: {password}\nEMAIL: {email}"
    
    def generate_email(self) -> str:
        """Generate fake email format"""
        subject = random.choice(self.email_subjects)
        starter = random.choice(self.email_starters)
        topic = random.choice(self.topics)
        
        message = f"{starter} the {topic}. Let me know if you need any additional information."
        
        return f"Subject: {subject}\nMessage: {message}"
    
    def generate_fake_plaintext(self) -> str:
        """Generate a fake plaintext using one of three strategies"""
        choice = random.randint(0, 2)
        
        if choice == 0:
            return self.generate_chat_message()
        elif choice == 1:
            return self.generate_credentials()
        else:
            return self.generate_email()
    
    def generate_fake_word(self) -> str:
        """Generate a single realistic word"""
        return random.choice(self.word_dictionary)


# Singleton instance
_fake_generator = None

def get_fake_generator() -> FakeGenerator:
    """Get the singleton FakeGenerator instance"""
    global _fake_generator
    if _fake_generator is None:
        _fake_generator = FakeGenerator()
    return _fake_generator


def generate_fake_plaintext() -> str:
    """Generate a fake plaintext"""
    return get_fake_generator().generate_fake_plaintext()


def generate_fake_word() -> str:
    """Generate a single fake word"""
    return get_fake_generator().generate_fake_word()
